zio-cache/shared/src/main/scala/zio/cache/Cache.scala

```diff
@@ -188,19 +190,96 @@ object Cache {
                     } else {
                       ZIO.done(exit)
                     }
+                  case MapValue.Refreshing(
+                        MapValue.Pending(_, promiseInProgress),
+                        MapValue.Complete(_, currentResult, _, ttl)
+                      ) =>
+                    trackAccess(key)
+                    trackHit()
+                    val now = Instant.now()
+                    if (now.isAfter(ttl)) {
+                      promiseInProgress.await
+                    } else {
+                      ZIO.done(currentResult)
+                    }
                 }
               }
+            }
 
+          override def refresh(k: Key): IO[Error, Value] =
+            ZIO.effectSuspendTotal {
+              var key: MapKey[Key]               = null
+              var promise: Promise[Error, Value] = null
+              var value                          = map.get(k)
+              if (value eq null) {
+                promise = Promise.unsafeMake[Error, Value](fiberId)
+                key = new MapKey(k)
+                value = map.putIfAbsent(k, MapValue.Pending(key, promise))
+              }
+              if (value eq null) {
+                retrieveFor(key, promise)
+              } else {
+                value match {
+                  case MapValue.Pending(_, promiseInProgress)                         =>
+                    promiseInProgress.await
+                  case completedResult @ MapValue.Complete(_, currentResult, _, ttl)  =>
+                    val now = Instant.now()
+                    if (now.isAfter(ttl)) {
+                      map.remove(k, value)
+                      get(k)
+                    } else {
+                      if (promise eq null) {
+                        promise = Promise.unsafeMake[Error, Value](fiberId)
+                      }
+                      if (key eq null) {
+                        key = new MapKey(k)
+                      }
+                      val refreshInProgress = MapValue.Refreshing(
+                        MapValue.Pending(key, promise),
+                        completedResult
+                      )
+                      map.put(k, refreshInProgress)
```

Comment:
We might want to compare `value` with what `map.put(k, refreshInProgress)` returns. If another thread might have updated the map with `MapValue.Complete` already so here we might trigger another re-fetch/computation. In terms of correctness, I think it's still correct but ideally I would like to eliminate the possibility of duplicating the work.

Reply:
Yes we need to use `replace` here to put the new value in the map only if the new value equals the old value.

Reply:
Fixed. Thanks Adam.